Fix bounds within buffer definition for MPI_ALL_TO_ALLV communication

rfj82982 opened the pull request:
Fix for issue #420

rfj82982 committed:
Fix bounds within buffer definition for MPI_ALL_TO_ALLV communication

pbartholomew08 reviewed (approved):
LGTM

mathrack commented:
LGTM.

After merging, this modification can be cherry-picked and merged inside `dev` and `main` as well.

Please update the Changelog.

rfj82982 committed:
Update CHANGELOG with MPI_ALL_TO_ALLV fix

Fixed the calculations of the displacement for MPI_ALL_TO_ALLV communication to avoid issue with 03 optimisation with some compilers.
